pkg/sidecar_controller/snapshot_controller.go

```diff
@@ -0,0 +1,579 @@
+	var err error
+
+	// Check if annotation exists
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnDeletionSecretRefName) && metav1.HasAnnotation(content.ObjectMeta, utils.AnnDeletionSecretRefNamespace) {
+		annDeletionSecretName := content.Annotations[utils.AnnDeletionSecretRefName]
+		annDeletionSecretNamespace := content.Annotations[utils.AnnDeletionSecretRefNamespace]
+
+		snapshotterSecretRef := &v1.SecretReference{}
+
+		if annDeletionSecretName == "" || annDeletionSecretNamespace == "" {
+			return nil, fmt.Errorf("cannot retrieve secrets for snapshot content %#v, err: secret name or namespace not specified", content.Name)
+		}
+
+		snapshotterSecretRef.Name = annDeletionSecretName
+		snapshotterSecretRef.Namespace = annDeletionSecretNamespace
+
+		snapshotterCredentials, err = utils.GetCredentials(ctrl.client, snapshotterSecretRef)
+		if err != nil {
+			// Continue with deletion, as the secret may have already been deleted.
+			klog.Errorf("Failed to get credentials for snapshot %s: %s", content.Name, err.Error())
+			return nil, fmt.Errorf("cannot get credentials for snapshot content %#v", content.Name)
+		}
+	}
+
+	return snapshotterCredentials, nil
+}
+
+// removeContentFinalizer removes a Finalizer for VolumeSnapshotContent.
+func (ctrl csiSnapshotSideCarController) removeContentFinalizer(content *crdv1.VolumeSnapshotContent) error {
+	contentClone := content.DeepCopy()
+	contentClone.ObjectMeta.Finalizers = slice.RemoveString(contentClone.ObjectMeta.Finalizers, utils.VolumeSnapshotContentFinalizer, nil)
+
+	_, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().Update(contentClone)
+	if err != nil {
+		return newControllerUpdateError(content.Name, err.Error())
+	}
+
+	_, err = ctrl.storeContentUpdate(contentClone)
+	if err != nil {
+		klog.Errorf("failed to update content store %v", err)
+	}
+
+	klog.V(5).Infof("Removed protection finalizer from volume snapshot content %s", content.Name)
+	return nil
+}
+
+// shouldDelete checks if content object should be deleted
+// if DeletionTimestamp is set on the content
+func (ctrl *csiSnapshotSideCarController) shouldDelete(content *crdv1.VolumeSnapshotContent) bool {
+	klog.V(5).Infof("Check if VolumeSnapshotContent[%s] should be deleted.", content.Name)
+
+	if content.ObjectMeta.DeletionTimestamp == nil {
+		return false
+	}
+	// 1) shouldDelete returns true if content is not bound
+	// (VolumeSnapshotRef.UID == "") for pre-provisioned snapshot
+	if content.Spec.Source.SnapshotHandle != nil && content.Spec.VolumeSnapshotRef.UID == "" {
+		return true
+	}
+	// 2) shouldDelete returns true if ShouldDelete annotation is set
+	// on the content with Delete policy.
+	// For content with Delet policy, we need to delete the physical
+	// snapshot resource first, and then remove content finalizer.
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnShouldDelete) {
+		return true
+	}
+	// 3) shouldDelete returns true for content with Retain policy.
```

Comment:
the last check may introduces problem and not necessary.
for example:
dynamically created snapshotcontent with retain policy = retain, user issues a delete via kubectl on VSC,
it should NOT delete the content BEFORE its corresponding snapshot has been deleted.
in a case where the VSC's bound VS has been deleted by user, common controller should have removed the finalizer on the VSC if retain policy is set to retain,
in this case, it will be simply an no-op in sidecar controller, the content will be deleted directly from the API server (unless there are other finalizers added via other controllers/user)

Reply:
>> in a case where the VSC's bound VS has been deleted by user, common controller should have removed the finalizer on the VSC if retain policy is set to retain,

Okay, that is fine, but then it behaves the same as we have in the current PR.  Since finalizer is already removed by the common controller, the content will be removed BEFORE its corresponding snapshot has been deleted.